File: lib/domain/book/book.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';

part 'book.freezed.dart';

/// Per-book reader settings persisted inside the Book `readConfig` record.
@freezed
class BookReadConfig with _$BookReadConfig {
  const BookReadConfig._();

  const factory BookReadConfig({
    @Default(false) bool reverseToc,
    @JsonKey(includeFromJson: false, includeToJson: false)
    @Default(<String, dynamic>{})
    Map<String, dynamic> extra,
  }) = _BookReadConfig;

  factory BookReadConfig.fromJson(Map<String, dynamic> json) {
    final map = Map<String, dynamic>.from(json);
    final rawReverse = map.remove('reverseToc');
    return BookReadConfig(
      reverseToc: rawReverse is bool ? rawReverse : false,
      extra: Map<String, dynamic>.unmodifiable(map),
    );
  }

  factory BookReadConfig.fromLegacyJson(
    dynamic value, {
    bool? legacyTopLevelReverseToc,
  }) {
    final map = value is Map
        ? Map<String, dynamic>.from(value)
        : <String, dynamic>{};
    final rawReverse = map.remove('reverseToc');
    final reverse = rawReverse is bool
        ? rawReverse
        : legacyTopLevelReverseToc ?? false;
    return BookReadConfig.fromJson({...map, 'reverseToc': reverse});
  }

  Map<String, dynamic> toJson() => {...extra, 'reverseToc': reverseToc};
}

/// 书籍领域实体，对应 Legado 的一本书。
@freezed
class Book with _$Book {
  const Book._();

  const factory Book({
    required String id,
    required String name,
    @Default('未知作者') String author,
    @Default('') String coverUrl,
    @Default('online') String type,
    @Default(0.0) double progress,
    String? currentChapter,
    String? lastChapter,
    @Default(0) int totalChapterNum,
    @Default(0) int durChapterIndex,
    @Default(0) int currentPageIndex,
    @Default(BookReadConfig()) BookReadConfig readConfig,
    @Default(false) bool isFavorite,
    @Default('') String sourceUrl,
    @Default('') String tocUrl,
    @Default('') String description,
    @Default('') String bookSourceUrl,
    @Default('') String group,
    @Default(0) int readIteration,
    @Default(false) bool simReadEnabled,
    @Default('') String simReadStartDate,
    @Default(0) int simReadStartChapter,
    @Default(3) int simReadDailyChapters,
    String? updatedAt,
  }) = _Book;

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id'] as String,
      name: json['name'] as String,
      author: json['author'] as String? ?? '未知作者',
      coverUrl: json['coverUrl'] as String? ?? '',
      type: json['type'] as String? ?? 'online',
      progress: (json['progress'] as num?)?.toDouble() ?? 0.0,
      currentChapter: json['currentChapter'] as String?,
      lastChapter: json['lastChapter'] as String?,
      totalChapterNum: (json['totalChapterNum'] as num?)?.toInt() ?? 0,
      durChapterIndex: (json['durChapterIndex'] as num?)?.toInt() ?? 0,
      currentPageIndex: (json['currentPageIndex'] as num?)?.toInt() ?? 0,
      readConfig: BookReadConfig.fromLegacyJson(
        json['readConfig'],
        legacyTopLevelReverseToc: json['reverseToc'] as bool?,
      ),
      isFavorite: json['isFavorite'] as bool? ?? false,
      sourceUrl: json['sourceUrl'] as String? ?? '',
      tocUrl: json['tocUrl'] as String? ?? '',
      description: json['description'] as String? ?? '',
      bookSourceUrl: json['bookSourceUrl'] as String? ?? '',
      group: json['group'] as String? ?? '',
      readIteration: (json['readIteration'] as num?)?.toInt() ?? 0,
      simReadEnabled: json['simReadEnabled'] as bool? ?? false,
      simReadStartDate: json['simReadStartDate'] as String? ?? '',
      simReadStartChapter: (json['simReadStartChapter'] as num?)?.toInt() ?? 0,
      simReadDailyChapters:
          ((json['simReadDailyChapters'] as num?)?.toInt() ?? 3).clamp(1, 999),
      updatedAt: json['updatedAt'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'author': author,
      'coverUrl': coverUrl,
      'type': type,
      'progress': progress,
      'currentChapter': currentChapter,
      'lastChapter': lastChapter,
      'totalChapterNum': totalChapterNum,
      'durChapterIndex': durChapterIndex,
      'currentPageIndex': currentPageIndex,
      'readConfig': readConfig.toJson(),
      'isFavorite': isFavorite,
      'sourceUrl': sourceUrl,
      'tocUrl': tocUrl,
      'description': description,
      'bookSourceUrl': bookSourceUrl,
      'group': group,
      'readIteration': readIteration,
      'simReadEnabled': simReadEnabled,
      'simReadStartDate': simReadStartDate,
      'simReadStartChapter': simReadStartChapter,
      'simReadDailyChapters': simReadDailyChapters,
      if (updatedAt != null) 'updatedAt': updatedAt,
    };
  }

  @override
  String toString() => 'Book($name - $author)';
}

File: rust/legado_engine/src/model/book.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Flutter/domain-facing projection of one persisted Legado book.
///
/// The JSON field names, nullable `currentChapter`, and empty `updatedAt`
/// fallback remain compatible with the existing `db_get_books` FFI result.
#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BookDto {
    pub id: String,
    pub name: String,
    pub author: String,
    pub cover_url: String,
    #[serde(rename = "type")]
    pub book_type: String,
    pub progress: f64,
    pub current_chapter: Option<String>,
    pub last_chapter: String,
    pub total_chapter_num: i64,
    pub dur_chapter_index: i64,
    pub current_page_index: i64,
    pub read_config: Value,
    pub is_favorite: bool,
    pub source_url: String,
    pub toc_url: String,
    pub description: String,
    pub book_source_url: String,
    pub group: String,
    pub read_iteration: i64,
    pub sim_read_enabled: bool,
    pub sim_read_start_date: String,
    pub sim_read_start_chapter: i64,
    pub sim_read_daily_chapters: i64,
    pub updated_at: String,
}

#[cfg(test)]
mod tests {
    use super::BookDto;
    use serde_json::json;

    #[test]
    fn book_dto_serializes_flutter_database_contract() {
        let dto = BookDto {
            id: "book-1".to_string(),
            name: "测试书".to_string(),
            author: "作者".to_string(),
            cover_url: "https://example.test/cover.jpg".to_string(),
            book_type: "online".to_string(),
            progress: 0.5,
            current_chapter: Some("第十章".to_string()),
            last_chapter: "第二十章".to_string(),
            total_chapter_num: 20,
            dur_chapter_index: 9,
            current_page_index: 123,
            read_config: json!({"reverseToc": true, "pageAnim": 2}),
            is_favorite: true,
            source_url: "https://example.test/book".to_string(),
            toc_url: "https://example.test/toc".to_string(),
            description: "简介".to_string(),
            book_source_url: "https://example.test/source".to_string(),
            group: "收藏".to_string(),
            read_iteration: 2,
            sim_read_enabled: true,
            sim_read_start_date: "2026-08-02".to_string(),
            sim_read_start_chapter: 3,
            sim_read_daily_chapters: 5,
            updated_at: "2026-08-02 12:00:00".to_string(),
        };

        let encoded = serde_json::to_value(dto).unwrap();

        assert_eq!(encoded["coverUrl"], "https://example.test/cover.jpg");
        assert_eq!(encoded["type"], "online");
        assert_eq!(encoded["currentChapter"], "第十章");
        assert_eq!(encoded["durChapterIndex"], 9);
        assert_eq!(encoded["currentPageIndex"], 123);
        assert_eq!(encoded["readConfig"]["reverseToc"], true);
        assert_eq!(encoded["readConfig"]["pageAnim"], 2);
        assert_eq!(encoded["simReadDailyChapters"], 5);
        assert_eq!(encoded["updatedAt"], "2026-08-02 12:00:00");
    }
}
